File: frontend/src/api/endpoints/signatureTemplates.ts
import { api } from "@/api/client";

export type SignatureTemplateFieldType =
  | "text"
  | "signature"
  | "initials"
  | "date"
  | "notary_stamp";

export interface SignatureTemplateField {
  id: string;
  type: SignatureTemplateFieldType;
  label: string;
  required: boolean;
}

export interface SignatureTemplateVersion {
  template_key: string;
  version: number;
  display_name: string;
  description: string;
  fields: SignatureTemplateField[];
  created_at: number;
  created_by: string;
  is_active: boolean;
}

export interface SignatureTemplateListResponse {
  templates: SignatureTemplateVersion[];
}

export interface SignatureTemplateVersionsResponse {
  template_key: string;
  versions: SignatureTemplateVersion[];
}

export interface CreateSignatureTemplateVersionInput {
  template_key: string;
  display_name: string;
  description?: string;
  fields: SignatureTemplateField[];
}

export interface SignatureTemplatePin {
  template_key: string;
  version: number;
}

export interface SignatureTemplateMigration {
  template_key: string;
  display_name: string;
  pinned_version: number;
  latest_version: number;
  needs_resigning: boolean;
}

export interface SignatureTemplateMigrationListResponse {
  migrations: SignatureTemplateMigration[];
}

export async function listSignatureTemplates(): Promise<SignatureTemplateListResponse> {
  return api.get<SignatureTemplateListResponse>("/ui/signing/templates");
}

export async function listSignatureTemplateVersions(
  templateKey: string,
): Promise<SignatureTemplateVersionsResponse> {
  return api.get<SignatureTemplateVersionsResponse>(
    `/ui/signing/templates/${encodeURIComponent(templateKey)}/versions`,
  );
}

export async function getSignatureTemplateVersion(
  templateKey: string,
  version: number,
): Promise<SignatureTemplateVersion> {
  return api.get<SignatureTemplateVersion>(
    `/ui/signing/templates/${encodeURIComponent(templateKey)}/versions/${version}`,
  );
}

export async function createSignatureTemplateVersion(
  input: CreateSignatureTemplateVersionInput,
): Promise<SignatureTemplateVersion> {
  return api.post<SignatureTemplateVersion>("/ui/signing/templates", input);
}

export async function checkSignatureTemplateMigration(
  pins: SignatureTemplatePin[],
): Promise<SignatureTemplateMigrationListResponse> {
  return api.post<SignatureTemplateMigrationListResponse>(
    "/ui/signing/templates/migration-check",
    { pins },
  );
}

File: app/routers/signature_templates.py
"""Versioned signature templates router (KYC-007).

Exposes CRUD-lite operations over versioned signature templates:
  - list latest template summaries
  - list all versions of a template
  - get a specific pinned version
  - create the next version (admin/notary only)
  - migration check: which pinned versions now have a newer template

Templates are versioned so that creating a packet pins a specific version;
publishing new terms creates a new version without mutating in-flight packets.
"""

from __future__ import annotations

from typing import Any, Dict

from fastapi import APIRouter, Depends, HTTPException

from app.auth.deps import AuthenticatedUser
from app.auth.policy import require_admin_or_root
from app.models import (
    CreateSignatureTemplateVersionIn,
    SignatureTemplateListOut,
    SignatureTemplateMigrationCheckIn,
    SignatureTemplateMigrationListOut,
    SignatureTemplateVersionOut,
    SignatureTemplateVersionsOut,
)
from app.services.sessions import require_ui_session
from app.services.signature_packet_flags import SignaturePdfDisabledError
from app.services.signature_template_versions import (
    TemplateVersionError,
    check_version_migration,
    create_template_version,
    get_template_version,
    list_all_template_keys,
    list_template_versions,
)

router = APIRouter(prefix="/ui/signing/templates", tags=["signature-templates"])


def _pdf_guard(exc: SignaturePdfDisabledError) -> HTTPException:
    return HTTPException(status_code=404, detail={"code": "signature_pdf_disabled"})


@router.get("", response_model=SignatureTemplateListOut)
def list_signature_templates(_ctx: Dict[str, str] = Depends(require_ui_session)) -> Dict[str, Any]:
    try:
        templates = list_all_template_keys()
    except SignaturePdfDisabledError as exc:
        raise _pdf_guard(exc) from exc
    return {"templates": templates}


@router.get("/{template_key}/versions", response_model=SignatureTemplateVersionsOut)
def list_signature_template_versions(
    template_key: str,
    _ctx: Dict[str, str] = Depends(require_ui_session),
) -> Dict[str, Any]:
    try:
        versions = list_template_versions(template_key)
    except SignaturePdfDisabledError as exc:
        raise _pdf_guard(exc) from exc
    return {"template_key": template_key, "versions": versions}


@router.get("/{template_key}/versions/{version}", response_model=SignatureTemplateVersionOut)
def get_signature_template_version(
    template_key: str,
    version: int,
    _ctx: Dict[str, str] = Depends(require_ui_session),
) -> Dict[str, Any]:
    try:
        out = get_template_version(template_key, version)
    except SignaturePdfDisabledError as exc:
        raise _pdf_guard(exc) from exc
    if not out:
        raise HTTPException(status_code=404, detail={"code": "signature_template_version_not_found"})
    return out


@router.post("", response_model=SignatureTemplateVersionOut)
def create_signature_template_version(
    body: CreateSignatureTemplateVersionIn,
    _ctx: Dict[str, str] = Depends(require_ui_session),
    user: AuthenticatedUser = Depends(require_admin_or_root),
) -> Dict[str, Any]:
    try:
        out = create_template_version(
            template_key=body.template_key,
            display_name=body.display_name,
            description=body.description,
            fields=[f.model_dump() for f in body.fields],
            created_by=user.sub,
        )
    except SignaturePdfDisabledError as exc:
        raise _pdf_guard(exc) from exc
    except TemplateVersionError as exc:
        raise HTTPException(status_code=400, detail={"code": str(exc)}) from exc
    return out


@router.post("/migration-check", response_model=SignatureTemplateMigrationListOut)
def signature_template_migration_check(
    body: SignatureTemplateMigrationCheckIn,
    _ctx: Dict[str, str] = Depends(require_ui_session),
) -> Dict[str, Any]:
    try:
        migrations = check_version_migration([p.model_dump() for p in body.pins])
    except SignaturePdfDisabledError as exc:
        raise _pdf_guard(exc) from exc
    return {"migrations": migrations}
